packages/blade/src/components/ActionList/ActionListBox.web.tsx

```diff
@@ -36,6 +44,120 @@ const _ActionListBox = React.forwardRef<HTMLDivElement, ActionListBoxProps>(
+const ActionListBox = assignWithoutSideEffects(React.memo(_ActionListBox), {
+  displayName: 'ActionListBox',
+});
 
-export { ActionListBox };
+const getVirtualItemParams = ({
+  theme,
+  isMobile,
+}: {
+  theme: Theme;
+  isMobile: boolean;
+}): {
+  itemHeight: number;
+  actionListBoxHeight: number;
+} => {
+  const itemHeightResponsive = getItemHeight(theme);
+  const actionListPadding = getActionListPadding(theme);
+  const actionListBoxHeight = actionListMaxHeight - actionListPadding * 2;
+
+  return {
+    itemHeight: isMobile
+      ? itemHeightResponsive.itemHeightMobile
+      : itemHeightResponsive.itemHeightDesktop,
+    actionListBoxHeight,
+  };
+};
+
+const useFilteredItems = (
+  children: React.ReactNode[],
+): {
+  itemData: React.ReactNode[];
+  itemCount: number;
+} => {
+  const childrenArray = React.Children.toArray(children); // Convert children to an array
+
+  const { filteredValues, hasAutoCompleteInBottomSheetHeader, dropdownTriggerer } = useDropdown();
+
+  const items = React.useMemo(() => {
+    const hasAutoComplete =
+      hasAutoCompleteInBottomSheetHeader ||
+      dropdownTriggerer === dropdownComponentIds.triggers.AutoComplete;
+
+    if (!hasAutoComplete) {
+      return childrenArray;
+    }
+
+    // @ts-expect-error: props does exist
+    const filteredItems = childrenArray.filter((item) => filteredValues.includes(item.props.value));
+    return filteredItems;
+  }, [filteredValues, hasAutoCompleteInBottomSheetHeader, dropdownTriggerer, childrenArray]);
+
+  return {
+    itemData: items,
+    itemCount: items.length,
+  };
+};
+
+const VirtualListItem = ({
+  index,
+  style,
+  data,
+}: {
+  index: number;
+  style: React.CSSProperties;
+  data: React.ReactNode[];
+}): React.ReactElement => {
+  return <div style={style}>{data[index]}</div>;
+};
+
+const _ActionListVirtualizedBox = React.forwardRef<HTMLDivElement, ActionListBoxProps>(
+  ({ childrenWithId, actionListItemWrapperRole, isMultiSelectable, ...rest }, ref) => {
+    const items = React.Children.toArray(childrenWithId); // Convert children to an array
+    const { isInBottomSheet } = useBottomSheetContext();
+    const { itemData, itemCount } = useFilteredItems(items);
+
+    const isMobile = useIsMobile();
+    const { theme } = useTheme();
+    const { itemHeight, actionListBoxHeight } = React.useMemo(
+      () => getVirtualItemParams({ theme, isMobile }),
+      // eslint-disable-next-line react-hooks/exhaustive-deps
+      [theme.name, isMobile],
+    );
+
+    return (
+      <StyledListBoxWrapper
+        isInBottomSheet={isInBottomSheet}
+        ref={ref}
+        {...makeAccessible({
+          role: actionListItemWrapperRole,
+          multiSelectable: actionListItemWrapperRole === 'listbox' ? isMultiSelectable : undefined,
+        })}
+        {...makeAnalyticsAttribute(rest)}
+      >
+        {itemCount < 30 ? (
+          childrenWithId
+        ) : (
+          <VirtualizedList
```

Comment:
maybe we can log an info that if item count is less switching to non virtualized list , might lead to a different experience in dev or prod env. if itemCount is less.

Reply:
Oh true good call. Will probably reduce the itemCount number and add a note in docs

Reply:
Added a note in ActionList docs props table. Will add better note when we write docs and tests and virtulization. Currently I have not added any documentation.

Reply:
> might lead to a different experience in dev or prod env. if itemCount is less.

Let's wrap it in __DEV__ flag, otherwise consumers may suddenly get lot of noise in their consoles.